phoenix-core/src/main/java/org/apache/phoenix/compile/WhereCompiler.java

```diff
@@ -364,7 +401,495 @@ public Void visit(KeyValueColumnExpression expression) {
+            case GREATER_OR_EQUAL:
+                return createComparisonExpression(LESS, lhs, rhs);
+            case GREATER:
+                return createComparisonExpression(LESS_OR_EQUAL, lhs, rhs);
+            default:
+                throw new IllegalArgumentException("Unexpected CompareOp of " + op);
+            }
+        }
+        private static List<Expression> negateChildren(List<Expression> children) {
+            List <Expression> list = new ArrayList<>(children.size());
+            for (Expression child : children) {
+                if (child instanceof ComparisonExpression) {
+                    list.add(negate((ComparisonExpression) child));
+                } else if (child instanceof OrExpression) {
+                    list.add(negate((OrExpression) child));
+                } else if (child instanceof AndExpression) {
+                    list.add(negate((AndExpression) child));
+                } else if (child instanceof ColumnExpression) {
+                    list.add(new NotExpression(child));
+                } else if (child instanceof NotExpression) {
+                    list.add(child.getChildren().get(0));
+                } else {
+                    throw new IllegalArgumentException("Unexpected Instance of " + child);
+                }
+            }
+            return list;
+        }
+        private static Expression negate(OrExpression node) {
+            return new AndExpression(negateChildren(node.getChildren()));
+        }
+
+        private static Expression negate(AndExpression node) {
+            return new OrExpression(negateChildren(node.getChildren()));
+        }
+        @Override public Expression visitLeave(NotExpression node, List<Expression> l) {
+            Expression child = l.get(0);
+            if (child instanceof OrExpression) {
+                return negate((OrExpression) child);
+            } else if (child instanceof AndExpression) {
+                return negate((AndExpression) child);
+            } else if (child instanceof ComparisonExpression) {
+                return negate((ComparisonExpression) child);
+            } else if (child instanceof NotExpression) {
+                return child.getChildren().get(0);
+            } else if (child instanceof IsNullExpression) {
+                return new IsNullExpression(ImmutableList.of(l.get(0).getChildren().get(0)),
+                        !((IsNullExpression) child).isNegate());
+            } else {
+                return new NotExpression(child);
+            }
+        }
+
+        private Expression transformInList(InListExpression node, boolean negate, List<Expression> l) {
+            List<Expression> list = new ArrayList<>(node.getKeyExpressions().size());
+            for (Expression element : node.getKeyExpressions()) {
+                if (negate) {
+                    list.add(createComparisonExpression(NOT_EQUAL, l.get(0), element));
```

Comment:
@kadirozde This doesn't seem to handle the case where the LHS of inlist can be a RVC expression like `(col1, col2) IN ((2,3), (7,8))`

Reply:
It does handle. Please see the new test for this in WhereComplierTest.